### PR TITLE
Optimize bins filtering by merging bins

### PR DESCRIPTION
**Pull request by fuzhaoyuan**
Describe changes proposed in this pull request:
- Merge numerical bins to reduce the number of bins passed to the database filtering

### CODE REVIEW

**Comment by haynescd** [comment]
@fuzhaoyuan Looks good. Just fix the one sonar issue and we are good

**Review by haynescd — APPROVED** [review]

File: src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java [hunk 1]
```diff
@@ -155,5 +162,61 @@ public static List<ClinicalDataCount> normalizeDataCounts(List<ClinicalDataCount
...
+                // merge adjacent numerical bins
+                else {
+                    BigDecimal start = dataFilterValue.getStart();
+                    BigDecimal end = dataFilterValue.getEnd();
+
+                    if (mergedStart == null && mergedEnd == null) {
+                        mergedStart = start;
+                        mergedEnd = end;
+                    }
+                    else if (mergedEnd.equals(start)) mergedEnd = end;
```
Comment (onursumer) [inline src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java:207]:
```suggestion
                    else if (mergedEnd.equals(start)) {
                        mergedEnd = end;
                    }
```

File: src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java [hunk 2]
```diff
@@ -155,5 +162,61 @@ public static List<ClinicalDataCount> normalizeDataCounts(List<ClinicalDataCount
...
+     * Merge the range of numerical bins in DataFilters to reduce the number of scans that runs on the database when filtering.
+     */
+    private static <T extends DataFilter> List<T> mergeDataFilters(List<T> filters) {
+        List<T> mergedDataFilters = new ArrayList<>();
+
+        for (T filter : filters) {
+            List<DataFilterValue> mergedValues = new ArrayList<>();
+
+            BigDecimal mergedStart = null, mergedEnd = null;
+            for (DataFilterValue dataFilterValue : filter.getValues()) {
```
Comment (onursumer) [inline src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java:193]: this logic only works if the bins are sorted properly, right? are we sure that bins are always sorted the way we want?

Comment (fuzhaoyuan) [reply src/main/java/org/cbioportal/service/util/StudyViewColumnarServiceUtil.java:193]: I had 2 assumptions: 1) they come in sorted and 2) they're continuous. I handled the discontinuous case by adjacent merging but if sorting is needed we'll need an extra part and way more combinations of possible cases. What would you suggest?

**Pull request merged**